frame_pose_tracker: Use frame names if they exist

EricCousineau-TRI opened the pull request:
This allows frame names to be used, such that frames (not just body centers) can be tracked.

EricCousineau-TRI commented:
+@rcory for feature review, please.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5Qx_hVfptlBav6qvlh:b-yvq8xq)*
<!-- Sent from Reviewable.io -->

rcory commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[manipulation/util/frame_pose_tracker.cc, line 78 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5V2H4_MbMImEkVY25-:-L5V2H4_MbMImEkVY250:byr6uw2) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/frame_pose_tracker.cc#L78)):*
> ```cpp
>     // TODO(eric.cousineau): Rather than use this nasty `try / catch` block,
>     // RigidBodyTree should automatically create a frame for each link, using
>     // an identity transform.
> ```

BTW, agreed.

---

*[manipulation/util/frame_pose_tracker.h, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5UnChvkI0u7ShhceUD:-L5UnChvkI0u7ShhceUE:b8e6isc) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/frame_pose_tracker.h#L66)):*
> ```objc
>    * RigidBodyFrame name (std::string) specifying which parent this frame
>    * should be attached to, and the model instance id in the @p tree that
>    * contains the corresponding frame.
> ```

BTW, "...corresponding *parent* frame".

---

*[manipulation/util/frame_pose_tracker.h, line 71 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5UqKqgwcYVTXKOry09:-L5UqKqgwcYVTXKOry0A:bd1rv6f) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/frame_pose_tracker.h#L71)):*
> ```objc
>    *        keeps a reference to this tree to calculate frame poses in the world
>    *        frame.
>    * @param frames_info A mapping from a unique frame or body name to a
> ```

I don't think this reads correctly. The key is always a unique frame name (and not a body name). Perhaps it should read something like

```...A mapping from a unique frame name to a std::pair consisting of the parent name and model instance id. The parent name can be either a body name or an existing frame name...```
```


---

*[manipulation/util/test/frame_pose_tracker_test.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5UxtOQIDJ5QpYVSLj_:-L5UxtOQIDJ5QpYVSLja:b-660qy8) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/test/frame_pose_tracker_test.cc#L47)):*
> ```cpp
>         std::make_pair("iiwa_link_1", model_id_);
>     frame_info_["iiwa_frame_2"] =
>         std::make_pair("iiwa_link_ee", model_id_);
> ```

These are all bodies. Would it make sense to add a frame parent to the testing mix? Looks like `iiwa14_polytope_collision.urdf` has a frame called `iiwa_frame_ee`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085)*
<!-- Sent from Reviewable.io -->

rcory commented:
Looks good. A few comments.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5V2YsJhR4pQbZdsYv3:b95remr)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[manipulation/util/frame_pose_tracker.cc, line 78 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5V2H4_MbMImEkVY25-:-L5V3kyMBkklAA9sETQd:b-84hhot) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/frame_pose_tracker.cc#L78)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, agreed.
</blockquote></details>

OK Kewl, made #8089

---

*[manipulation/util/frame_pose_tracker.h, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5UnChvkI0u7ShhceUD:-L5V53pTYzYfP5udkm5G:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/frame_pose_tracker.h#L66)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, "...corresponding *parent* frame".
</blockquote></details>

Done.

---

*[manipulation/util/frame_pose_tracker.h, line 71 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5UqKqgwcYVTXKOry09:-L5V54Px2N5HlyNskRCq:b-1ukrtd) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/frame_pose_tracker.h#L71)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

I don't think this reads correctly. The key is always a unique frame name (and not a body name). Perhaps it should read something like

```...A mapping from a unique frame name to a std::pair consisting of the parent name and model instance id. The parent name can be either a body name or an existing frame name...```
</blockquote></details>

Done. Tried a slight adjustment to your wording. Does this seem accurate? (I'm a bit hesitant to say `unique frame name` from the perspective of `FramePoseTracker`, because really it seems like it can be any arbitrary name, as long as it's unique within this mapping.)

---

*[manipulation/util/test/frame_pose_tracker_test.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5UxtOQIDJ5QpYVSLj_:-L5V5Gff9CcBVWxA3czs:bhy6tdn) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/test/frame_pose_tracker_test.cc#L47)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

These are all bodies. Would it make sense to add a frame parent to the testing mix? Looks like `iiwa14_polytope_collision.urdf` has a frame called `iiwa_frame_ee`?
</blockquote></details>

Done. D'oh, had meant to use a frame - thanks for catching that!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085)*
<!-- Sent from Reviewable.io -->

rcory commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[manipulation/util/frame_pose_tracker.h, line 71 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5UqKqgwcYVTXKOry09:-L5VbXzIv8j-c4GCAXI3:bkoxnzz) ([raw file](https://github.com/robotlocomotion/drake/blob/7d4bf65af51ba8dea349c479fc99b804adc1539e/manipulation/util/frame_pose_tracker.h#L71)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done. Tried a slight adjustment to your wording. Does this seem accurate? (I'm a bit hesitant to say `unique frame name` from the perspective of `FramePoseTracker`, because really it seems like it can be any arbitrary name, as long as it's unique within this mapping.)
</blockquote></details>

Cool. I'm good with that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5Vc0MkFvGnd0NhCUQP:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rcory commented:
+@sherm1 for platform review.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5VcQgIlVKVjQJFDdG8:bfvx2sg)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform pass done. This depends on uniqueness of body and frame names -- how do we know that is true? Also, I think there is an easy enough way to avoid the non-compliant try/catch blocks, PTAL.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[manipulation/util/frame_pose_tracker.cc, line 84 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WPur-7QCveO6l08az:-L5WPur-7QCveO6l08b-:b9hlmy7) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/frame_pose_tracker.cc#L84)):*
> ```cpp
>       parent_body = parent_frame->get_mutable_rigid_body();
>       X_BF = parent_frame->get_transform_to_body();
>     } catch (const std::logic_error&) {
> ```

try/catch isn't allowed by the styleguide and I don't think this qualifies as an exception. Please consider adding a `tree->MaybeFindFrame()` method or something like that and use it here (I'm assuming it would be an easy copy-and-hack of `tree->findFrame()`). #8089 is a more substantial proposal involving default frames.

---

*[manipulation/util/frame_pose_tracker.h, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WR7LkilEUL96OgMBX:-L5WR7LkilEUL96OgMBY:b-z943jp) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/frame_pose_tracker.h#L71)):*
> ```objc
>    *        keeps a reference to this tree to calculate frame poses in the world
>    *        frame.
>    * @param frames_info A mapping from a name (unique in this mapping) to a
> ```

BTW how do we know that body names and frame names are disjoint? Is that guaranteed somewhere?

---

*[manipulation/util/test/frame_pose_tracker_test.cc, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WOw2s9UfDng011cFk:-L5WOw2tqrbg6-YoO-J4:b4bjq0f) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/test/frame_pose_tracker_test.cc#L61)):*
> ```cpp
>       std::tie(name, model_id) = frame_info.second;
>       RigidBody<double>* parent_body = nullptr;
>       Eigen::Isometry3d T_BF;  // From parent frame to body
> ```

BTW nit: our [standard notation](http://drake.mit.edu/doxygen_cxx/group__multibody__spatial__pose.html) for transforms is "X_BF" rather than "T_BF".

---

*[manipulation/util/test/frame_pose_tracker_test.cc, line 62 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WROf4Ye5V5WG9maiR:-L5WROf4Ye5V5WG9maiS:b-r7b3vy) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/test/frame_pose_tracker_test.cc#L62)):*
> ```cpp
>       RigidBody<double>* parent_body = nullptr;
>       Eigen::Isometry3d T_BF;  // From parent frame to body
>       // @note See TODO in `frame_pose_tracker` about `try / catch` block.
> ```

BTW see my comment there.

---

*[manipulation/util/test/frame_pose_tracker_test.cc, line 100 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WRWqwLLEfZkw5Jmc3:-L5WRWqwLLEfZkw5Jmc4:b-rw4t98) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/test/frame_pose_tracker_test.cc#L100)):*
> ```cpp
>   std::unique_ptr<RigidBodyTree<double>> tree_;
>   int model_id_;
>   Eigen::Isometry3d T_PF_;  // From frame to parent (frame or body)
> ```

FYI we generally frown on the "from / to" description for rotations and transforms because it is ambiguous unless you have a particular usage in mind (e.g., application of the transform on the left to a point on the right). If you want you could say:
```
// X_PF is the transform giving the pose of the F frame in the P frame.
// Given a position vector `p_FQ` of a point Q measured from and expressed in F,
// `p_PQ=X_PF*p_FQ` give the position vector of that same point Q measured
// from and expressed in P.
```
or you could just link to the [documentation](http://drake.mit.edu/doxygen_cxx/group__multibody__spatial__pose.html).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5WTUABDfIIVv7HbJgX:b-3wcqeg)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Went for the slightly more (but not too much more) involved method of sealing off `bodies` and `frames` from `RigidBodyTree`, prevent duplicates (though our mixed multi-map-ish usage of `model_name` and `model_instance_id` kinda makes it messy), and adds a default frame for the body.

Also provides cleaner guarantees of non-duplicates (at construction time, vs. just query time).

---

Review status: 1 of 37 files reviewed at latest revision, 5 unresolved discussions.

---

*[manipulation/util/frame_pose_tracker.cc, line 84 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WPur-7QCveO6l08az:-L5XTr-2bJtgNZICaK7d:b-4iamo7) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/frame_pose_tracker.cc#L84)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

try/catch isn't allowed by the styleguide and I don't think this qualifies as an exception. Please consider adding a `tree->MaybeFindFrame()` method or something like that and use it here (I'm assuming it would be an easy copy-and-hack of `tree->findFrame()`). #8089 is a more substantial proposal involving default frames.
</blockquote></details>

Done. Removed, and went ahead and plugged through #8089. Thankfully wasn't too involved.

(Incorporated in this PR - commits are curated.)

---

*[manipulation/util/frame_pose_tracker.h, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WR7LkilEUL96OgMBX:-L5XU4r76XONr_8vv5G4:bvdorze) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/frame_pose_tracker.h#L71)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW how do we know that body names and frame names are disjoint? Is that guaranteed somewhere?
</blockquote></details>

Done. Addressed by addressing #8089. You are guaranteed to have a frame that mirrors a body, but you should never have a duplicate frame or body.

---

*[manipulation/util/test/frame_pose_tracker_test.cc, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WOw2s9UfDng011cFk:-L5XWDW-d6Nfm_ALjgBy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/test/frame_pose_tracker_test.cc#L61)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW nit: our [standard notation](http://drake.mit.edu/doxygen_cxx/group__multibody__spatial__pose.html) for transforms is "X_BF" rather than "T_BF".
</blockquote></details>

Done.

---

*[manipulation/util/test/frame_pose_tracker_test.cc, line 62 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WROf4Ye5V5WG9maiR:-L5XWDn8lLbPX2LYl3MH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/test/frame_pose_tracker_test.cc#L62)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW see my comment there.
</blockquote></details>

Done.

---

*[manipulation/util/test/frame_pose_tracker_test.cc, line 100 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5WRWqwLLEfZkw5Jmc3:-L5XWEL9lMddbpOLCAbP:b-63ym9y) ([raw file](https://github.com/robotlocomotion/drake/blob/a649e398460df8bde6d1c6c9f3a57daa46e4f6e2/manipulation/util/test/frame_pose_tracker_test.cc#L100)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI we generally frown on the "from / to" description for rotations and transforms because it is ambiguous unless you have a particular usage in mind (e.g., application of the transform on the left to a point on the right). If you want you could say:
```
// X_PF is the transform giving the pose of the F frame in the P frame.
// Given a position vector `p_FQ` of a point Q measured from and expressed in F,
// `p_PQ=X_PF*p_FQ` give the position vector of that same point Q measured
// from and expressed in P.
```
or you could just link to the [documentation](http://drake.mit.edu/doxygen_cxx/group__multibody__spatial__pose.html).
</blockquote></details>

Done. Went for a shorter version - the main thing I want to do is tersely explain `P`. Is this OK?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5XWUkQeoWk-n_t77xw:b-ni3ffq)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
- [ ] If/when this PR lands with `RigidBodyTree` modifications, I'll submit a PR to spartan to squash usages of `bodies` and `frames` (looks relatively sparse).

sherm1 commented:
So nice to see the accessors instead of direct member access! Thanks for doing that, Eric.
A few minor issues and some BTWs & FYIs, PTAL. Otherwise, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 35 of 38 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[multibody/rigid_body_tree.cc, line 3020 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_bnORSPy2HdTO55kp:-L5_bnORSPy2HdTO55kq:bbh25cb) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3020)):*
> ```cpp
> 
> template <typename T>
> void RigidBodyTree<T>::addFrame(std::shared_ptr<RigidBodyFrame<T>> frame) {
> ```

FYI would this be a good time to rename this `AddFrame()`?

---

*[multibody/rigid_body_tree.cc, line 3036 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_aPRf0tPQJD1kf3tU:-L5_aPRf0tPQJD1kf3tV:b-8g8og2) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3036)):*
> ```cpp
>                 "Frame '{}', with model instance id {} ({}), already "
>                 "registered!",
>                 frame->get_name(), model_name, frame->get_model_instance_id()));
> ```

BTW the last two arguments are reversed from the message's expectations.

---

*[multibody/rigid_body_tree.cc, line 3041 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_acQhwdJtI3lfYFrx:-L5_acQhwdJtI3lfYFry:b-i60j8k) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3041)):*
> ```cpp
>   }
> 
>   // TODO(eric.cousineau): Throw error if an instance has the same frame.
> ```

BTW not sure what you mean here -- didn't you just do that check and throw above?

---

*[multibody/rigid_body_tree.cc, line 3048 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_c05nxxIMFvrhlyzw:-L5_c05nxxIMFvrhlyzx:b8v0loc) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3048)):*
> ```cpp
> 
> template <typename T>
> RigidBody<T>* RigidBodyTree<T>::add_rigid_body(
> ```

BTW would this be a good time to rename this `AddRigidBody()`?

---

*[multibody/rigid_body_tree.cc, line 3060 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_aooWH1zwpGQCop-s:-L5_aooWH1zwpGQCop-t:b-1pv8wb) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3060)):*
> ```cpp
>                 "Body '{}', with model instance id {} ({}), already "
>                 "registered!",
>                 body->get_name(), model_name, body->get_model_instance_id()));
> ```

BTW arguments are reversed from message.

---

*[multibody/rigid_body_tree.h, line 1 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_bBdbOGgl7fGPEqsd:-L5_bBdbOGgl7fGPEqse:b-sj48sf) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.h#L1)):*
> ```objc
> #pragma once
> ```

BTW the new policies enforced by addFrame() and add_rigid_body() should be doxygen-documented somewhere in the public interface.

---

*[multibody/rigid_body_tree.h, line 1500 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5__t7D8iCdPknq34oK:-L5__t7D8iCdPknq34oL:bc83q0s) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.h#L1500)):*
> ```objc
> 
>   // Rigid body objects
>   // TODO(amcastro-tri): make private and start using accessors body(int).
> ```

BTW can remove this TODO now.

---

*[multibody/rigid_body_tree.h, line 1502 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_cs1quzXTEIU1OTbb:-L5_cs1quzXTEIU1OTbc:bjruvod) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.h#L1502)):*
> ```objc
>   // TODO(amcastro-tri): make private and start using accessors body(int).
>   // TODO(amcastro-tri): rename to bodies_ to follow Google's style guide once.
>   // accessors are used throughout the code.
> ```

BTW since these are now private, please consider renaming them bodies_ and frames_ and removing this TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5_dRM6aDChP-ALirfH:b-fod9fk)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
This PR changes the public API of RigidBodyTree.  Please consider deprecating the old API for a while, before removing it.

EricCousineau-TRI commented:
Welcome!

> This PR changes the public API of RigidBodyTree.  Please consider deprecating the old API for a while, before removing it.

Done. I'm deprecating read access, and removing writable access, since that ruins the encapsulation of `add_rigid_body` and `addFrame`, and does not seem to be used in downstream code.

---

Review status: 35 of 38 files reviewed at latest revision, 8 unresolved discussions.

---

*[multibody/rigid_body_tree.cc, line 3020 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_bnORSPy2HdTO55kp:-L5jSB5opWrzTWcleUUe:bbj6kvo) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3020)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI would this be a good time to rename this `AddFrame()`?
</blockquote></details>

OK I'd prefer to defer (added a TODO). This is an aesthetic / non-functional change, while the other changes are motivated by functionality (assuring encapsulation).

---

*[multibody/rigid_body_tree.cc, line 3036 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_aPRf0tPQJD1kf3tU:-L5jSlsdsVba4WCL2Uar:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3036)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW the last two arguments are reversed from the message's expectations.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.cc, line 3041 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_acQhwdJtI3lfYFrx:-L5jSuomMxJDxOU6ED5e:b-drn8ya) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3041)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW not sure what you mean here -- didn't you just do that check and throw above?
</blockquote></details>

Done. Oops! Forgot to remove that.

---

*[multibody/rigid_body_tree.cc, line 3048 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_c05nxxIMFvrhlyzw:-L5jSWWn_QLkwyqJtpMe:b5v2cy) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3048)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW would this be a good time to rename this `AddRigidBody()`?
</blockquote></details>

OK I'd prefer to defer. Same reasoning as `AddFrame`.

---

*[multibody/rigid_body_tree.cc, line 3060 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_aooWH1zwpGQCop-s:-L5jS_acusYtA2AyOD1P:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.cc#L3060)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW arguments are reversed from message.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 1 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_bBdbOGgl7fGPEqsd:-L5jRXciYS3VtTBTcOoo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.h#L1)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW the new policies enforced by addFrame() and add_rigid_body() should be doxygen-documented somewhere in the public interface.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 1500 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5__t7D8iCdPknq34oK:-L5jRYkphzQxVhbxoJQi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.h#L1500)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW can remove this TODO now.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 1502 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5_cs1quzXTEIU1OTbb:-L5jR_1VS_HRi-ovNwnd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3d33defd65487f2e72c00a3445f981b30c62fba0/multibody/rigid_body_tree.h#L1502)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW since these are now private, please consider renaming them bodies_ and frames_ and removing this TODO.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5jRfe-YyHbiS0aJGv6:b-ahhopo)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Looks great, one new (minor) BTW.

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[multibody/rigid_body_tree.h, line 1412 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5jWH2ZXeHQcInVjigZ:-L5jWH2ZXeHQcInVjig_:b-sp20mk) ([raw file](https://github.com/robotlocomotion/drake/blob/ea859b8a8a4cd2bf74090b48094db7a0e8fbb50e/multibody/rigid_body_tree.h#L1412)):*
> ```objc
>    * @return A bare, unowned pointer to the @p body.
>    * @pre A body with the same identifying information (name and model id /
>    * name) should not already exist in the tree.
> ```

BTW I think there also can't be a frame already there with the same name as the proposed new body?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-:-L5jWqFuEkEFqiGLej3G:b4fjnlj)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 36 of 38 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/rigid_body_tree.h, line 1412 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8085#-L5jWH2ZXeHQcInVjigZ:-L5l6dvd4-dCqgntjTvN:b-3h8a35) ([raw file](https://github.com/robotlocomotion/drake/blob/ea859b8a8a4cd2bf74090b48094db7a0e8fbb50e/multibody/rigid_body_tree.h#L1412)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I think there also can't be a frame already there with the same name as the proposed new body?
</blockquote></details>

Done. Good point, missed that!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8085)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please
